File: backend/src/test/java/edonymyeon/backend/member/domain/NicknameTest.java
package edonymyeon.backend.member.domain;

import static edonymyeon.backend.global.exception.ExceptionInformation.*;
import static edonymyeon.backend.member.domain.SocialInfo.SocialType;
import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.DisplayNameGenerator.ReplaceUnderscores;

import edonymyeon.backend.global.exception.EdonymyeonException;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.Test;

@DisplayNameGeneration(ReplaceUnderscores.class)
@SuppressWarnings("NonAsciiCharacters")
class NicknameTest {

    @Test
    void 닉네임을_String으로_생성한다() {
        //given
        final String nickname = "newNickname";

        //when
        //then
        Assertions.assertDoesNotThrow(() -> Nickname.from(nickname));
    }

    @Test
    void 닉네임을_SocialType으로_생성한다() {
        //given
        final SocialType socialType = SocialType.KAKAO;

        //when
        //then
        Assertions.assertDoesNotThrow(() -> Nickname.from(socialType));
    }

    @Test
    void 닉네임이_20글자를_초과하면_예외가_발생한다() {
        //given
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < 21; i++) {
            stringBuilder.append(i);
        }
        final String nickname = stringBuilder.toString();

        //when
        //then
        assertThatThrownBy(() -> Nickname.from(nickname))
                .isInstanceOf(EdonymyeonException.class)
                .hasMessage(MEMBER_NICKNAME_INVALID.getMessage());
    }
}

File: android/domain/src/test/kotlin/com/domain/edonymyeon/model/NicknameTest.kt
package com.domain.edonymyeon.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows

internal class NicknameTest {
    @Test
    fun `닉네임은 0글자 이상이다`() {
        assertDoesNotThrow {
            Nickname.create("가")
        }
    }

    @Test
    fun `닉네임이 0글자이면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Nickname.create("")
        }
    }

    @Test
    fun `닉네임에 공백을 넣으면 제거된다`() {
        val nickname = Nickname.create("가 ")
        assertThat(nickname.value).isEqualTo("가")
    }

    @Test
    fun `닉네임 검증에서 0글자가 넘으면 true가 반환된다`() {
        val result = Nickname.validate("가")
        assertThat(result).isEqualTo(true)
    }

    @Test
    fun `닉네임 검증에서 0글자가 넘지 않으면 false가 반환된다`() {
        val result = Nickname.validate("")
        assertThat(result).isEqualTo(false)
    }
}
